Docs multiplugin page-wide tabs merge (#11461)
openvinotoolkit/openvino: docs/OV_Runtime_UG/multi_device.md

# Running on multiple devices simultaneously {#openvino_docs_OV_UG_Running_on_multiple_devices}

> The Multi-Device execution mode, or MULTI for short, uses a "virtual" or a "proxy" device, which does not bind to a specific type of hardware. Instead, it assigns available computing devices to particular inference requests, which are then executed in parallel. 

Reviewer: "uses" sounds a bit misleading, either "constitutes", or maybe "acts"

Author: nice change, done.

> The Multi-Device execution mode, or MULTI for short, uses a "virtual" or a "proxy" device, which does not bind to a specific type of hardware. Instead, it assigns available computing devices to particular inference requests, which are then executed in parallel. 

Author: ```suggestion
The Multi-Device execution mode, or MULTI for short, acts as a "virtual" or a "proxy" device, which does not bind to a specific type of hardware. Instead, it assigns available computing devices to particular inference requests, which are then executed in parallel. 
```

The potential gains from using Multi-Device execution are:
* improved throughput from using multiple devices at once,
* increase in performance stability due to multiple devices sharing inference workload.

> What is important, the Multi-Device mode does not change tha application logic, so it does not require you to explicitly compile the model on every device or create and balance inference requests. From the application point of view, this is just another plugin that handles the actual machinery.

Reviewer: "just another plugin that handles " - > "just a conventional **device** that **internally** handles the actual machinery"

Author: sounds better and more formal, gladly changing... done.

> What is important, the Multi-Device mode does not change tha application logic, so it does not require you to explicitly compile the model on every device or create and balance inference requests. From the application point of view, this is just another plugin that handles the actual machinery.

Author: ```suggestion
What is important, the Multi-Device mode does not change the application logic, so it does not require you to explicitly compile the model on every device or create and balance inference requests. From the application point of view, this is just a conventional device that internally handles the actual machinery.
```

> What is important, the Multi-Device mode does not change the application logic, so it does not require you to explicitly compile the model on every device or create and balance inference requests. From the application point of view, this is just a conventional device that internally handles the actual machinery.

Reviewer 2: ```suggestion
Importantly, the Multi-Device mode does not change the application logic, so it does not require you to explicitly compile the model on every device or create and balance inference requests. To the application, this is just a conventional device that internally handles the actual devices.
```

Author: ```suggestion
Importantly, the Multi-Device mode does not change the application logic, so it does not require you to explicitly compile the model on every device or create and balance inference requests. It appears to use a typical device but internally handles the actual hardware.
```

> Note that performance increase in this mode comes from utilizing multiple devices at once. This means that you need to provide the devices with enough inference requests to keep them busy or you will not benefit much from using MULTI.

Reviewer 2: ```suggestion
Note that the performance increase in this mode comes from utilizing multiple devices at once. This means that you need to provide the devices with enough inference requests to keep them busy, otherwise you will not benefit much from using MULTI.
```

## Using the Multi-Device Mode 

Following the OpenVINO™ naming convention, the Multi-Device mode is assigned the label of “MULTI.” The only configuration option available for it is a prioritized list of devices to use:

@sphinxdirective

+---------------------------+---------------------------------+-----------------------------------------------------------+
| Property                  | Property values                 | Description                                               |
+===========================+=================================+===========================================================+

> | <device candidate list>   | | AUTO: <device names>          || Specifies the devices available for selection.           |

Reviewer: MULTI:

Author: done

> | <device candidate list>   | | AUTO: <device names>          || Specifies the devices available for selection.           |

Reviewer: unlike AUTO, these are NOT candidates: if loading to any device from the list fails, the entire MULTI would fail as well

> | <device candidate list>   | | AUTO: <device names>          || Specifies the devices available for selection.           |

Author: ```suggestion
| <device candidate list>   | | MULTI: <device names>         || Specifies the devices available for selection.           |
```

> | <device candidate list>   | | MULTI: <device names>         || Specifies the devices available for selection.           |

Author: ```suggestion
| <device list>             | | MULTI: <device names>         || Specifies the devices available for selection.           |
```

|                           | | comma-separated, no spaces    || The device sequence will be taken as priority            |
+---------------------------+---------------------------------+| from high to low.                                        |
| ov::device::priorities    | | device names                  || Priorities can be set directly as a string.              |
|                           | | comma-separated, no spaces    |                                                           |
+---------------------------+---------------------------------+-----------------------------------------------------------+

@endsphinxdirective

> The device candidate list allows users to customize the priority and limit the choice of devices available to MULTI. If not specified, the plugin assumes all the devices present in the system can be used. Importantly, the list may also specify the number of requests for MULTI to keep for each device, as described below.

Reviewer: " If not specified, the plugin assumes all the devices present in the system can be used"- this is not true, unlike the AUTO, the MULTI does require the explicit list of the devices

Reviewer: pls see above, the list is not "candidates".

Author: Oh, ok, that is an important difference. Repairing it now.

> The device candidate list allows users to customize the priority and limit the choice of devices available to MULTI. If not specified, the plugin assumes all the devices present in the system can be used. Importantly, the list may also specify the number of requests for MULTI to keep for each device, as described below.

Author: ```suggestion
Specifying the device list explicitly is required by MULTI, as it defines the devices available for inference and sets their priorities.  Importantly, the list may also specify the number of requests for MULTI to keep for each device, as described below.
```

> Note, that OpenVINO™ Runtime lets you use “GPU” as an alias for “GPU.0” in function calls. More detail on enumerating devices can be found in [Working with devices](supported_plugins/Device_Plugins.md).

Reviewer 2: ```suggestion
Note that OpenVINO™ Runtime lets you use “GPU” as an alias for “GPU.0” in function calls. More details on enumerating devices can be found in [Working with devices](supported_plugins/Device_Plugins.md).
```

       :language: cpp
       :fragment: [part0]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_multi.py
       :language: python
       :fragment: [MULTI_0]

@endsphinxdirective

> Notice that MULTI allows you to **change device priorities on the fly**. You can alter the order, exclude a device, and bring an excluded device back. Still, it does not allow to add new devices.

Reviewer 2: ```suggestion
Notice that MULTI allows you to **change device priorities on the fly**. You can alter the order, exclude a device, and bring an excluded device back. Still, it does not allow adding new devices.
```

       :language: cpp
       :fragment: [part1]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_multi.py
       :language: python
       :fragment: [MULTI_1]

@endsphinxdirective

> One more thing you can define is the **number of requests to allocate for each device**. You can do it simply by adding the number to each device in parentheses, like so: `"MULTI:CPU(2),GPU(2)"`. Although, this method is not recommended as it is not performance-portable. The suggested approach is to configure individual devices and query the resulting number of requests to be used at the application level, as described below, in [Configuring Individual Devices and Creating MULTI On Top](#configuring-the-individual-devices-and-creating-the-multi-device-on-top).

Reviewer 2: ```suggestion
One more thing you can define is the **number of requests to allocate for each device**. You can do it simply by adding the number to each device in parentheses, like this: `"MULTI:CPU(2),GPU(2)"`. However, this method is not recommended as it is not performance-portable. The suggested approach is to configure individual devices and query the resulting number of requests to be used at the application level, as described in [Configuring Individual Devices and Creating MULTI On Top](#configuring-the-individual-devices-and-creating-the-multi-device-on-top).
```

> To check what devices are present in the system, you can use the Device API. For information on how to do it, check [Query device properties and configuration](supported_plugins/config_properties.md)

Reviewer 2: ```suggestion
To check what devices are present in the system, you can use the Device API. For information on how to do it, check [Query device properties and configuration](supported_plugins/config_properties.md).
```

@endsphinxdirective

## Using the Multi-Device with OpenVINO Samples and Benchmarking Performance

> To see how the Multi-Device execution is used in practice and test its performance, take a look at OpenVINO™ samples. All samples supporting the `-d` command-line option (which stands for “device”) will accept the MULTI mode natively. The Benchmark Application will be a perfect place to start – it presents the optimal performance of the plugin without the need for additional settings, like the number of requests or CPU threads. 

Author: ```suggestion
To see how the Multi-Device execution is used in practice and test its performance, take a look at OpenVINO's Benchmark Application which presents the optimal performance of the plugin without the need for additional settings, like the number of requests or CPU threads. 
```

> To see how the Multi-Device execution is used in practice and test its performance, take a look at OpenVINO™ samples. All samples supporting the `-d` command-line option (which stands for “device”) will accept the plugin out-of-the-box. The Benchmark Application will be a perfect place to start – it presents the optimal performance of the plugin without the need for additional settings, like the number of requests or CPU threads. 

Reviewer: "will accept the plugin out-of-the-box" is confusing... we are usually talking about the "devices", rather than "plugins".
the benchmark_app is today the on;ly option, as the rest of the samples are API examples, and the MULTI is about  the performance, whihc only the benchamrk_app reports.

Author: I guess "will accept the MULTI mode natively" is a better choice of words. Is it ok? 
But does it mean we should not mention the '-d" option? If so, we can remove the whole sentence.

> To see how the Multi-Device execution is used in practice and test its performance, take a look at OpenVINO™ samples. All samples supporting the `-d` command-line option (which stands for “device”) will accept the plugin out-of-the-box. The Benchmark Application will be a perfect place to start – it presents the optimal performance of the plugin without the need for additional settings, like the number of requests or CPU threads. 

Author: ```suggestion
To see how the Multi-Device execution is used in practice and test its performance, take a look at OpenVINO™ samples. All samples supporting the `-d` command-line option (which stands for “device”) will accept the MULTI mode natively. The Benchmark Application will be a perfect place to start – it presents the optimal performance of the plugin without the need for additional settings, like the number of requests or CPU threads. 
```

Here is an example command to evaluate performance of HDDL+GPU: 

```sh
./benchmark_app –d MULTI:HDDL,GPU –m <model> -i <input> -niter 1000
```

For more information, refer to the [C++](../../samples/cpp/benchmark_app/README.md) or [Python](../../tools/benchmark_tool/README.md) version instructions.	

@sphinxdirective
.. note::

>    You can use the FP16 IR to work with auto-device.

Author: ```suggestion
   You can keep using the FP16 IR, without converting it to FP32, to work with MULTI, even if some of the listed devices do not support it. The conversion will be done automatically for you.
```

>    You can use the FP16 IR to work with auto-device.

Reviewer: MULTI

Reviewer: the message is that no need to use the FP32 IR when any device in the list doesn't suppurt the FP16, the conversion is automacally happens by the OV

Author: The whole point reworded, I hope now it's more clear
